### PR TITLE
Automated abi generation

### LINKED ISSUE #197: STAT-168 ⁃ Automated ABI generation for smart contract
Currently developers need to manually write the ABI (JSON) for their application, this is error prone when all the data should be available through the contract code.

Acceptance test criteria for the ABI Generator:
https://github.com/elmato/abi-generator-atc

### PR DESCRIPTION
**Pull request by elmato**
Resolves #197

### CODE REVIEW

**Comment by bytemaster** [comment]
This pull request is 3 weeks old, is it still valid and where can I read the documentation on how to use this?

**Comment by elmato** [comment]
yes, it's still valid!
#197 has a comment with the "documentation", i will move it to an .md file

**Comment by bytemaster** [comment]
Please resolve conflicts.

File: CMakeModules/wasm.cmake [hunk 1]
```diff
@@ -121,9 +122,23 @@ macro(add_wast_target target INCLUDE_FOLDERS DESTINATION_FOLDER)
...
+  #   DEPENDS ${target}.s abi_generator
+  #   COMMAND abi_generator -extra-arg=-c -extra-arg=--std=c++14 -extra-arg=--target=wasm32
+  #             -extra-arg=-I${INCLUDE_FOLDERS} -extra-arg=-fparse-all-comments
+  #             -destination-file=${DESTINATION_FOLDER}/${target}.abi -verbose=1 -context=${target} ${SOURCE_FILES} --             
+  #   COMMENT "Generating ABI ${target}.abi"
+  #   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+  #   VERBATIM
+  # )
+  # set_property(DIRECTORY APPEND PROPERTY ADDITIONAL_MAKE_CLEAN_FILES ${DESTINATION_FOLDER}/${target}.abi)
+
```
Comment (heifner) [inline CMakeModules/wasm.cmake]: Please add a comment why this commented out section is here.

File: contracts/eoslib/raw.hpp [hunk 2]
```diff
@@ -100,6 +103,34 @@ namespace eos {
       s.skip(size.value);
     }
 
+    // FixedString32
```
Comment (heifner) [inline contracts/eoslib/raw.hpp]: Please add full doxygen comments for all methods.

File: libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+      if( UNLIKELY(!(TEST)) ) \
+      {                                                                      \
+        if( fc::enable_record_assert_trip )                                  \
+           fc::record_assert_trip( __FILE__, __LINE__, #TEST );              \
+        FC_THROW_EXCEPTION( eos::abi_generator::abi_generation_exception, #TEST ": "  __VA_ARGS__ ); \
+      }                                                                      \
+    FC_MULTILINE_MACRO_END \
+  )
+
+struct AbiGenerator {
```
Comment (heifner) [inline libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp]: Please add doxygen comments for struct and all methods.

File: libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+      rawText = smatch.suffix();
+    } 
+
+  } FC_CAPTURE_AND_RETHROW() }
+
+  bool is64bit(const string& type) {
+    types::AbiSerializer abis(*output);
+    return abis.isInteger(type) && abis.getIntegerSize(type) == 64;
+    //return type_size[type] == 64;
+  }
```
Comment (heifner) [inline libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp]: Here and elsewhere please remove commented out code.

File: libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+  bool                   verbose;
+  int                    optimizations;
+  types::Abi*            output;
+  clang::ASTContext*     context;
+  CompilerInstance*      compiler_instance;
+  //map<string, uint64_t>  type_size;
+  map<string, string>    full_types;
+  string                 abi_context;
+
+  AbiGenerator() : optimizations(0), output(nullptr), context(nullptr), compiler_instance(nullptr) {
```
Comment (heifner) [inline libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp]: FYI. New coding standard is:
* All structs, classes, methods, types should be lower case with _ between words including constants.
* All Template arguments should be CamelCase
* All MACROS should be all UPPERCASE

However, this was all created before that was made clear.

We do need doxygen for all classes and public methods.

**Comment by elmato** [comment]
Acceptance test criteria for the ABI Generator
https://github.com/elmato/abi-generator-atc

File: contracts/currency/currency.abi [hunk 6]
```diff
@@ -5,15 +5,15 @@
...
+      "name": "Transfer",
       "base": "",
       "fields": {
         "from": "account_name",
         "to": "account_name",
         "amount": "uint64"
       }
     },{
-      "name": "account",
+      "name": "Account",
```
Comment (heifner) [inline contracts/currency/currency.abi]: Should Account and Transfer be lowercase to match struct in currency.hpp?

**Pull request merged**